src/core/tsi/alts_transport_security.cc

```diff
@@ -24,53 +24,52 @@
 
 static alts_shared_resource g_alts_resource;
 
-static void grpc_tsi_alts_wait_for_cq_drain() {
-  while (!g_alts_resource.is_cq_drained) {
+static void wait_for_resource_destroy() {
+  gpr_mu_lock(&g_alts_resource.mu);
+  while (!g_alts_resource.is_resource_destroyed) {
     gpr_cv_wait(&g_alts_resource.cv, &g_alts_resource.mu,
                 gpr_inf_future(GPR_CLOCK_REALTIME));
   }
+  gpr_mu_unlock(&g_alts_resource.mu);
 }
 
-gpr_mu* grpc_tsi_alts_shared_resource_get_mu() { return &g_alts_resource.mu; }
-
-void grpc_tsi_alts_signal_for_cq_destroy() {
-  gpr_mu_lock(&g_alts_resource.mu);
-  g_alts_resource.is_cq_drained = true;
+static void signal_for_resource_destroy_locked() {
+  g_alts_resource.is_resource_destroyed = true;
   gpr_cv_signal(&g_alts_resource.cv);
-  gpr_mu_unlock(&g_alts_resource.mu);
 }
 
 void grpc_tsi_alts_init() {
   g_alts_resource.channel = nullptr;
   g_alts_resource.cq = nullptr;
-  g_alts_resource.is_cq_drained = false;
+  g_alts_resource.is_resource_destroyed = false;
   gpr_mu_init(&g_alts_resource.mu);
   gpr_cv_init(&g_alts_resource.cv);
   gpr_ref_init(&g_alts_resource.refcount, 0);
 }
 
 void grpc_tsi_alts_shutdown() {
+  wait_for_resource_destroy();
   GPR_ASSERT(g_alts_resource.cq == nullptr);
   gpr_cv_destroy(&g_alts_resource.cv);
   gpr_mu_destroy(&g_alts_resource.mu);
 }
 
-alts_shared_resource* grpc_tsi_alts_shared_resource_ref() {
+alts_shared_resource* grpc_tsi_alts_shared_resource_create() {
   gpr_mu_lock(&g_alts_resource.mu);
   gpr_ref(&g_alts_resource.refcount);
   gpr_mu_unlock(&g_alts_resource.mu);
   return &g_alts_resource;
 }
 
-void grpc_tsi_alts_shared_resource_unref(alts_shared_resource* resource) {
+void grpc_tsi_alts_shared_resource_destroy(alts_shared_resource* resource) {
   gpr_mu_lock(&resource->mu);
   if (gpr_unref(&resource->refcount) && resource->cq != nullptr) {
     grpc_completion_queue_shutdown(resource->cq);
-    grpc_tsi_alts_wait_for_cq_drain();
     grpc_completion_queue_destroy(resource->cq);
     resource->cq = nullptr;
     grpc_channel_destroy(resource->channel);
     resource->thread.Join();
   }
+  signal_for_resource_destroy_locked();
```

Comment:
Shouldn't this go inside of the block above, so that it only executes when the refcount goes to zero?

Reply:
There could be a situation when refcount is zero but resource->cq is null (e.g., only credential but no channel), and I think we need to take into account that case as well.

Reply:
Ah I understand what you mean now. The function should be placed inside the `if` branch. Thanks for pointing it out.